src/main/java/io/reactivex/internal/operators/flowable/FlowableBufferBoundarySupplier.java

```diff
@@ -196,24 +196,20 @@ void next() {
 
             BufferBoundarySubscriber<T, U, B> bs = new BufferBoundarySubscriber<T, U, B>(this);
 
-            Disposable o = other.get();
-
-            if (!other.compareAndSet(o, bs)) {
-                return;
-            }
-
-            U b;
-            synchronized (this) {
-                b = buffer;
-                if (b == null) {
-                    return;
+            if (DisposableHelper.replace(other, bs)) {
```

Comment:
Use dedicated helper instead of a CAS, plus the old code could unset a terminal indicator.